### PR TITLE
Support for CancellationTokenSource in TestContext to help in timeout…

### PR DESCRIPTION
**Pull request by cltshivash**
Support for CancellationTokenSource in TestContext to help in timeout scenario

### CODE REVIEW

**Comment by acesiddhu** [comment]
do we have a RFC for this change?

**Comment by cltshivash** [comment]
> do we have a RFC for this change?

Yes will author one.

File: src/TestFramework/Extension.Shared/TestContext.cs [hunk 1]
```diff
@@ -21,6 +22,11 @@ public abstract class TestContext
         /// </summary>
         public abstract IDictionary Properties { get; }
 
+        /// <summary>
+        /// Gets or sets the cancellation token source that is cancelled when test timesout or when cancelled the test will be aborted
```
Comment (singhsarab) [inline src/TestFramework/Extension.Shared/TestContext.cs]: Nit: Gets or sets the cancellation token source. It is cancelled when the test times out. When cancelled, the test gets aborted.

Just a suggestion, please fix the above string as suited. public surface so being careful.

File: test/UnitTests/PlatformServices.Desktop.Unit.Tests/Services/DesktopThreadOperationsTests.cs [hunk 2]
```diff
@@ -100,5 +105,33 @@ public void ExecuteWithAbortSafetyShouldCatchThreadAbortExceptionsAndResetAbort(
...
+        public void TokenCancelShouldAbortExecutingAction()
+        {
+            // setup
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            // act
+            cancellationTokenSource.CancelAfter(100);
+            var result = this.asyncOperations.Execute(() => { Thread.Sleep(10000); }, 100000, cancellationTokenSource.Token);
+
+            // validate
```
Comment (singhsarab) [inline test/UnitTests/PlatformServices.Desktop.Unit.Tests/Services/DesktopThreadOperationsTests.cs:119]: Shouldn't we validate for something True as well ?

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs [hunk 3]
```diff
@@ -692,16 +694,18 @@ private TestResult ExecuteInternalWithTimeout(object[] arguments)
...
-
-                // If the method times out, then
-                //
-                // 1. If the test is stuck, then we can get CannotUnloadAppDomain exception.
-                //
-                // Which are handled as follows: -
-                //
-                // For #1, we are now restarting the execution process if adapter fails to unload app-domain.
+                // Timed out or canceled
                 string errorMessage = string.Format(CultureInfo.CurrentCulture, Resource.Execution_Test_Timeout, this.TestMethodName);
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:697]: Can we add tests for this change.

Comment (cltshivash) [reply src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:697]: Will do

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs [hunk 4]
```diff
@@ -692,16 +694,18 @@ private TestResult ExecuteInternalWithTimeout(object[] arguments)
...
+                if (this.TestMethodOptions.TestContext.Context.CancellationTokenSource.IsCancellationRequested)
+                {
+                    errorMessage = string.Format(CultureInfo.CurrentCulture, Resource.Execution_Test_Cancelled, this.TestMethodName);
+                }
+                else
+                {
+                    // Cancel the token source
```
Comment (jayaranigarg) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs]: Nit: Cancel the token source as it has timed-out

File: src/Adapter/PlatformServices.Desktop/Services/DesktopThreadOperations.cs [hunk 5]
```diff
@@ -75,7 +84,39 @@ public void ExecuteWithAbortSafety(Action action)
...
 
+        private static bool JoinThread(int timeout, Thread executionThread)
+        {
+            try
+            {
+                return executionThread.Join(timeout);
+            }
+            catch (ThreadStateException)
+            {
+                // Catch and discard ThreadStateException. If Abort is called on a thread that has been suspended,
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Services/DesktopThreadOperations.cs]: Nit: Comments are not related to **JoinThread**

File: src/Adapter/PlatformServices.Shared/netstandard1.0/Services/ns10TestContextImplementation.cs [hunk 6]
```diff
@@ -61,6 +61,7 @@ public TestContextImplementation(ITestMethod testMethod, StringWriter writer, ID
             this.testMethod = testMethod;
             this.properties = new Dictionary<string, object>(properties);
             this.stringWriter = writer;
+            this.CancellationTokenSource = new CancellationTokenSource();
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Shared/netstandard1.0/Services/ns10TestContextImplementation.cs:64]: Add this change parallely for DesktopTestContextImplementation as well : https://github.com/Microsoft/testfx/blob/master/src/Adapter/PlatformServices.Desktop/Services/DesktopTestContextImplementation.cs#L76

File: src/TestFramework/Extension.Shared/TestContext.cs [hunk 7]
```diff
@@ -21,6 +22,11 @@ public abstract class TestContext
...
+        /// </summary>
+        public CancellationTokenSource CancellationTokenSource { get; protected set; }
```
Comment (jayaranigarg) [inline src/TestFramework/Extension.Shared/TestContext.cs]: Add this change for Desktop TestContext as well : https://github.com/Microsoft/testfx/blob/master/src/TestFramework/Extension.Desktop/TestContext.cs#L18

File: test/UnitTests/PlatformServices.Shared.Unit.Tests/netstandard1.0/ns10ThreadOperationsTests.cs [hunk 8]
```diff
@@ -35,26 +35,36 @@ public ThreadOperationsTests()
...
         public void ExecuteShouldReturnFalseIftheActionTimesout()
         {
-            Action action = () => { Task.Delay(100).Wait(); };
+            void action()
+            {
+                Task.Delay(100).Wait();
+            }
 
-            Assert.IsFalse(this.asyncOperations.Execute(action, 1));
+            CancellationTokenSource tokenSource = new CancellationTokenSource();
```
Comment (jayaranigarg) [inline test/UnitTests/PlatformServices.Shared.Unit.Tests/netstandard1.0/ns10ThreadOperationsTests.cs:56]: We should add a test where Execute returns false when the cancellationToken is cancelled.

**Comment by jayaranigarg** [comment]
I would encourage you to add E2E test for the scenario which shows how can user leverage this CancellationToken.

File: src/TestFramework/Extension.Shared/TestContext.cs (hunk 1 above)
Comment (cltshivash) [reply src/TestFramework/Extension.Shared/TestContext.cs]: Done.

File: src/Adapter/PlatformServices.Desktop/Services/DesktopThreadOperations.cs (hunk 5 above)
Comment (cltshivash) [reply src/Adapter/PlatformServices.Desktop/Services/DesktopThreadOperations.cs]: Will fix it.

File: src/Adapter/PlatformServices.Shared/netstandard1.0/Services/ns10TestContextImplementation.cs (hunk 6 above)
Comment (cltshivash) [reply src/Adapter/PlatformServices.Shared/netstandard1.0/Services/ns10TestContextImplementation.cs:64]: Done

File: test/UnitTests/PlatformServices.Shared.Unit.Tests/netstandard1.0/ns10ThreadOperationsTests.cs (hunk 8 above)
Comment (cltshivash) [reply test/UnitTests/PlatformServices.Shared.Unit.Tests/netstandard1.0/ns10ThreadOperationsTests.cs:56]: Done. Added.

File: test/UnitTests/PlatformServices.Desktop.Unit.Tests/Services/DesktopThreadOperationsTests.cs (hunk 2 above)
Comment (cltshivash) [reply test/UnitTests/PlatformServices.Desktop.Unit.Tests/Services/DesktopThreadOperationsTests.cs:119]: Done.

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs (hunk 4 above)
Comment (cltshivash) [reply src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs]: Done.

**Comment by cltshivash** [comment]
> I would encourage you to add E2E test for the scenario which shows how can user leverage this CancellationToken.

Added E2E tests

File: src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx [hunk 9]
```diff
@@ -317,4 +317,7 @@ Error: {1}</value>
   <data name="UTA_TestMethodExpectedParameters" xml:space="preserve">
     <value>Only data driven test methods can have parameters. Did you intend to use [DataRow] or [DynamicData]?</value>
   </data>
+  <data name="Execution_Test_Cancelled" xml:space="preserve">
```
Comment (singhsarab) [inline src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx:320]: Please xlf file changes associated with this change

File: test/E2ETests/TestAssets/TimeoutTestProject/TerimnateLongRunningTasksUsingTokenTestClass.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace TimeoutTestProject
+{
+    [TestClass]
+    public class TerimnateLongRunningTasksUsingTokenTestClass
```
Comment (singhsarab) [inline test/E2ETests/TestAssets/TimeoutTestProject/TerimnateLongRunningTasksUsingTokenTestClass.cs:13]: >TerimnateLongRunningTasksUsingTokenTestClass [](start = 17, length = 44)

Nit: *Terminate*

File: src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx (hunk 9 above)
Comment (jayaranigarg) [reply src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx:320]: Run command "build.cmd -uxlf". You would have to commit the xlf files changes post running the script.

**Comment by singhsarab** [comment]
@cltshivash Can you please add an RFC for this. If added already, please add the link here. Thanks.

**Pull request merged**